loopy/transform/merge_arrays.py

```diff
@@ -0,0 +1,113 @@
+__copyright__ = "Copyright (C) 2022 Isuru Fernando"
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+__doc__ = """
+.. currentmodule:: loopy
+.. autofunction:: merge_temporary_arrays
+"""
+
+from loopy.kernel.data import auto
+from loopy.symbolic import SubstitutionRuleMappingContext
+from loopy.kernel import LoopKernel
+from loopy.translation_unit import for_each_kernel
+
+import pymbolic.primitives as prim
+from pytools import all_equal
+
+from typing import Text, Sequence, Optional, Union
+
+
+@for_each_kernel
+def merge_temporary_arrays(
+        kernel: LoopKernel,
+        array_names: Union[Text, Sequence[Text]],
+        new_name: Optional[Text] = None,
+        axis_nr: int = 0) -> LoopKernel:
+    """Merges temporary arrays into one array along the axis given by *axis_nr*.
+
+    :arg array_names: a list of names of temporary variables or arguments. May
+        also be a comma-separated string of these.
+
+    :arg axis_nr: the (zero-based) index of the axis of the arrays to be merged.
+
+    :arg new_name: new name for the merged temporary. If not given, the name
+        of the first array given by *array_names* is used.
+    """
+    assert isinstance(kernel, LoopKernel)
+
+    if isinstance(array_names, str):
+        array_names = [i.strip() for i in array_names.split(",") if i.strip()]
+
+    new_name = new_name or array_names[0]
+    new_aggregate = prim.Variable(new_name)
+
+    tvs = []
+    offsets = {}
+    count = 0
+    for array_name in array_names:
+        tv = kernel.temporary_variables[array_name]
+        if tv.shape in [None, auto]:
+            raise ValueError(f"Shape of temporary variable '{array_name}' is "
+                    "unknown. Cannot merge with unknown shapes")
+        offsets[array_name] = count
+        count += tv.shape[axis_nr]
+
+        shape = list(tv.shape)
+        shape[axis_nr] = 1
```

Comment:
The logic here is a bit stateful. Maybe separate the sanity check logic from computing the new shape? (IMO, pytato.concatenate does a cleaner job here, but YMMV :))

Reply:
You mean move line 74 and 75 to a new loop?

Reply:
Yep!

Reply:
Done in 5bbbd116524af9c5941fb133f24d67cdf49b1a42